anomaly_detector/core/job.py

```diff
@@ -0,0 +1,102 @@
+from abc import ABCMeta, abstractmethod
+import os
+from prometheus_client import Counter
+from anomaly_detector.model import W2VModel
+from anomaly_detector.adapters import FeedbackStrategy, SomModelAdapter, SomStorageAdapter
+from anomaly_detector.exception import EmptyDataSetException, ModelSaveException, ModelLoadException
+import time
+from opentracing_instrumentation.request_context import get_current_span, span_in_context
+
+TRAINING_COUNT = Counter("aiops_lad_train_count", "count of training runs")
+INFER_COUNT = Counter("aiops_lad_inference_count", "count of inference runs")
+
+
+class AbstractCommand(metaclass=ABCMeta):
+    """Generic command that gets executed when added to the TaskManager steps."""
+
+    @abstractmethod
+    def execute(self):
+        """Generic method that gets executed by the task manager."""
+        raise NotImplementedError("Please implement the <execute method>")
+
+
+class SomTrainJob(AbstractCommand):
+    """Som Training logic custom for each model. For this model we need to setup some configs."""
+
+    recreate_models = False
+
+    def __init__(self, node_map=24, model_adapter=None, recreate_model=True):
+        """Initializing properties for training."""
+        self.node_map = node_map
+        self.model_adapter = model_adapter
+        self.recreate_model = recreate_model
+
+    def execute_with_tracing(self, tracer):
+        """Train models for anomaly detection with tracing enabled."""
+        with tracer.start_span('aiops_train_execute', child_of=get_current_span()) as span:
+            with span_in_context(span):
+                return self.execute()
+
+    def execute(self):
+        """Train models for anomaly detection."""
+        TRAINING_COUNT.inc()
+        dataframe, raw_data = self.model_adapter.preprocess(config_type="train",
+                                                            recreate_model=self.recreate_model)
+        if not raw_data:
+            raise EmptyDataSetException("no new logs found.")
+        # After first time training we will only update w2v model not recreate it everytime.
+        dist = self.model_adapter.train(node_map=self.node_map, data=dataframe, recreate_model=self.recreate_model)
+        self.recreate_model = False
+        return 0, dist
+
+
+class SomInferenceJob(AbstractCommand):
+    """Som Inference implementation."""
+
+    def __init__(self, model_adapter=None, sleep=True, recreate_model=False):
+        """Initialize inference command."""
+        self.model_adapter = model_adapter
+        self.sleep = sleep
+        self.recreate_model = recreate_model
+
+    def execute_with_tracing(self, tracer):
+        """Will retrain with fresh data and perform predictions in batches."""
+        with tracer.start_span('aiops_inference_execute', child_of=get_current_span()) as span:
+            with span_in_context(span):
+                return self.execute()
+
+    def execute(self):
+        """Execute inference steps and loop."""
```

Comment:
What's the difference between `execute` and `execute_with_tracing`?
Because the `SomTrainJob` methods have different doc strings

Reply:
Execute with tracing provides jeagar tracing to allow us to measure how long the functions take.

Reply:
Fixed.